instrumentation/messaging/src/main/java/brave/messaging/MessagingAdapter.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2013-2019 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.messaging;
+
+import brave.Span;
+import brave.internal.Nullable;
+
+/**
+ * @param <Chan> the type of the channel
+ * @param <Msg> the type of the message
+ * @param <C> the type that carriers the trace context, usually headers
+ */
+// abstract class instead of interface to allow method adds before Java 1.8
+public abstract class MessagingAdapter<Chan, Msg, C> {
+  // TODO: make some of these methods not abstract as they don't have meaning for all impls
+  // jeqo: The only one that fit this description is correlationId. JMS and AMQP use it, but Kafka and RocketMQ don't. RocketMQ has a concept of transactionId that might replace it, but then I'd say each impl can add it's own additional identifiers/offsets.
+
+  /** Returns the trace context carrier from the message. Usually, this is headers. */
+  public abstract C carrier(Msg message);
+
+  /**
+   * Messaging channel, e.g. kafka queue or JMS topic name. {@code null} if unreadable.
+   *
+   * <p>Conventionally associated with the key "message.channel"
+   */
+  @Nullable public abstract String channel(Chan channel);
```

Comment:
Should we replace `channel` to `channel_name`, as we have `channel_kind`? Similar to `span_name`, `service_name`, etc.